Feature/questions user flow

Marin988 opened the pull request.

Marin988 committed:
Add Confluence client
Add QuestionAnswer model and State model
Add SearchesProcessor
Add UserFlowProcessor

Marin988 committed:
Implement GetInitialQuestions function
Add the UserFlow plugin
Add frontend questions module and simple page

Marin988 committed:
Add functions for getting questions and saving questions

Marin988 committed:
Using mat tree with adding functionality
Add edit funcitonlity
Subquestions are now stored as json
Add title to question model

Marin988 committed:
Convert nested lists of questions into one layered list to save them in db
Userflow processor logic
Add property boolean array to QuestionAnswer, representing the MentorMaterType they are valid for

Marin988 committed:
frontend design
add traits to questionAnswer entity
create questionAnswerViewModel
changes to state entity
userflow tests

Marin988 committed:
remove unused storage tasks
remove unused question answer type enum
remove unused css classes
remove unused component functions
use bootstrap to make the page look better

Marin988 committed:
remove unnecessary attributes from questionAnswer view model
remove confluence client and searches processor

rosenkolev commented on src/MentorBot.Functions/Processors/UserFlow/UserFlowProcessor.cs:
```diff
@@ -0,0 +1,160 @@
+                return new ChatEventResult("No UserFlow hosts are configured.");
+            }
+
+            var user = hosts.GetValue<string>(UserFlowProperties.User);
```
Use `originalChatEvent.Message.Sender`

rosenkolev commented on src/MentorBot.Functions/Processors/UserFlow/UserFlowProcessor.cs:
```diff
@@ -0,0 +1,160 @@
+
+            var questions = await _storageService.GetAllQuestionsAsync();
+
+            var relativeQuestions = questions.ToList();
```
Why the `.ToList()`

rosenkolev commented on src/MentorBot.Functions/Processors/UserFlow/UserFlowProcessor.cs:
```diff
@@ -0,0 +1,160 @@
+            {
+                relativeQuestions = relativeQuestions.Where(q => q.Parents == null || q.Parents.Count == 0).OrderBy(q => q.Index).ToList();
+                var mentorMaterTypesCard = CreateCard(relativeQuestions, "Select your Mentor Mater Type.");
+                state.Active = true;
```
Why not set `CurrentQuestionId` and `ShownQuestionIds`

rosenkolev commented on src/MentorBot.Functions/Processors/UserFlow/UserFlowProcessor.cs:
```diff
@@ -0,0 +1,160 @@
+            else
+            {
+                relativeQuestions = questions
+                .Where(q => q.RequiredTraits.Any(t => state.Traits.FirstOrDefault(st => st == t) != null)
```
Duplicated code here

rosenkolev commented on src/MentorBot.Functions/Processors/UserFlow/UserFlowProcessor.cs:
```diff
@@ -0,0 +1,160 @@
+
+            if (string.IsNullOrEmpty(parentId) || parentId == "null")
+            {
+                relativeQuestions = relativeQuestions.Where(q => q.Parents == null || q.Parents.Count == 0).ToList();
```
Duplicated code here

Marin988 committed:
get the user email of the message sender

rosenkolev commented on src/MentorBot.WebLearningCenter/ClientApp/package.json:
```diff
@@ -19,14 +19,15 @@
     "@angular/compiler": "13.1.3",
     "@angular/core": "13.1.3",
     "@angular/forms": "13.1.3",
+    "@angular/material": "^13.2.4",
```
Why use both angular material and bootstrap?

Marin988 committed:
Add delete question functionallity
Add deletion confirmation
Update tests

Marin988 committed:
Fix push errors
Add functionallity to remove traits from the "Added" and "Removed" lists

Marin988 committed:
Add input constraints and error messages
Remove unneeded code from "get-quesitons" in Queries

rosenkolev commented on src/MentorBot.Functions/AzureFunctions/Commands.cs:
```diff
@@ -85,6 +87,115 @@ public static class Commands
+
+                await storageService.DeleteQuestionAnswerAsync(question);
+
+                foreach (var child in children)
```
Separate into methods.

rosenkolev commented on src/MentorBot.Functions/AzureFunctions/Queries.cs:
```diff
@@ -188,6 +189,69 @@ public static class Queries
+
+            var questionsToDelete = new List<QuestionAnswerViewModel>();
+
+            foreach (var question in questionViewModels)
```
In separate method

rosenkolev commented on src/MentorBot.Functions/Models/ViewModels/QuestionAnswerViewModel.cs:
```diff
@@ -0,0 +1,38 @@
+        public bool IsAnswer { get; set; }
+
+        /// <summary> Gets or sets a value indicating whether the entity is edited. </summary>
+        public bool IsEdited { get; set; }
```
remove isEdited

rosenkolev commented on src/MentorBot.WebLearningCenter/ClientApp/src/app/questions/components/question-page/question-page.component.html:
```diff
@@ -0,0 +1,193 @@
+<link href="https://fonts.googleapis.com/icon?family=Material+Icons" rel="stylesheet">
```
remove from here.

rosenkolev commented on src/MentorBot.WebLearningCenter/ClientApp/src/app/questions/components/question-page/question-page.component.scss:
```diff
@@ -0,0 +1,139 @@
+
+.delete-traits {
+  background-color: red;
+  // opacity: 0.5;
```
remove comment

rosenkolev commented on src/MentorBot.WebLearningCenter/ClientApp/src/app/questions/components/question-page/question-page.component.html:
```diff
@@ -0,0 +1,193 @@
+
+</div>
+
+<!-- Copyright 2018 Google Inc. All Rights Reserved.
```
Remove this comment.
Make the file with less html

rosenkolev commented on src/MentorBot.WebLearningCenter/ClientApp/src/app/questions/components/question-page/question-page.component.ts:
```diff
@@ -0,0 +1,435 @@
+import { QuestionService } from '../../question.service';
+
+/** Flat to-do item node with expandable and level information */
+export class TodoItemFlatNode {
```
move model to another file.
may be `app/questions/questions.models.ts`

rosenkolev commented on src/MentorBot.WebLearningCenter/ClientApp/src/app/questions/components/question-page/question-page.component.ts:
```diff
@@ -0,0 +1,435 @@
+ * If a node is a category, it has children items and new items can be added under the category.
+ */
+@Injectable()
+export class ChecklistDatabase {
```
Move this class to another file.

rosenkolev replied:
lets implement `DataSource<Question>`

rosenkolev commented on src/MentorBot.WebLearningCenter/ClientApp/src/app/questions/components/question-page/question-page.component.ts:
```diff
@@ -0,0 +1,435 @@
+  }
+
+  initialize() {
+    this._questionService.getQuestions().subscribe(questions => this.dataChange.next(questions));
```
Where is the unsubscribe?
var subscription = getQuestions().subscribe(() => {
...
subscription.unsibscribe()
})

`
 getQuestions().pipe(take(1)).subscribe(() => {
 
 });
 `

rosenkolev commented on src/MentorBot.WebLearningCenter/ClientApp/src/app/questions/components/question-page/question-page.component.ts:
```diff
@@ -0,0 +1,435 @@
+
+  /** Add an item to to-do list */
+  insertItem(isAnswer: boolean, requiredTraits: string[], acquireTraits: string[], parents: string[], name: string, content: string) {
+    var node = {
```
do not use var

rosenkolev replied:
do not use `var`

rosenkolev commented on src/MentorBot.WebLearningCenter/ClientApp/src/app/questions/components/question-page/question-page.component.ts:
```diff
@@ -0,0 +1,435 @@
+   * Build the file structure tree. The `value` is the Json object, or a sub-tree of a Json object.
+   * The return value is the list of `TodoItemNode`.
+   */
+  buildFileTree(obj: object, level: number): Question[] {
```
Record<string, Question[]> as type.
Do not use object or any

rosenkolev commented on src/MentorBot.WebLearningCenter/ClientApp/src/app/questions/components/question-page/question-page.component.ts:
```diff
@@ -0,0 +1,435 @@
+        node.parents[element] = 'new parent';
+      });
+
+      this.dataChange.value.push(node);
```
Why is value.push?

rosenkolev commented on src/MentorBot.WebLearningCenter/ClientApp/src/app/questions/components/question-page/question-page.component.ts:
```diff
@@ -0,0 +1,435 @@
+  }
+
+  insertTopItem(isAnswer: boolean, requiredTraits: string[], acquireTraits: string[], name: string, content: string) {
+    this.dataChange.value.push({
```
`
this.dataChange.next([
   ...this.data,
   {
     test
   }
]);
`

rosenkolev replied:
this.dataChange.pipe(pairwise(previous, current) => {})

rosenkolev commented on src/MentorBot.WebLearningCenter/ClientApp/src/app/questions/components/question-page/question-page.component.ts:
```diff
@@ -0,0 +1,435 @@
+      node.requiredTraits.splice(deleteIndex, 1);
+    });
+
+    this.dataChange.next(this.data);
```
const newNode = {
  ...node,
  title: name,
  
}

rosenkolev commented on src/MentorBot.WebLearningCenter/ClientApp/src/app/questions/components/question-page/question-page.component.ts:
```diff
@@ -0,0 +1,435 @@
+    this.treeControl = new FlatTreeControl<TodoItemFlatNode>(this.getLevel, this.isExpandable);
+    this.dataSource = new MatTreeFlatDataSource(this.treeControl, this.treeFlattener);
+
+    database.dataChange.subscribe(data => {
```
no unsubscribe

rosenkolev commented on src/MentorBot.WebLearningCenter/ClientApp/src/app/questions/components/question-page/question-page.component.ts:
```diff
@@ -0,0 +1,435 @@
+      addTraits[i].remove();
+      i--;
+    }
+    this.editedNode = {} as TodoItemFlatNode;
```
const emptyNode = {
  title: '',
  requiredTraits: [],
  parentNodes: []
};
this.editedNode = { ...emptyNode  };
```
<div *ngFor="let trait in requiredTraits" />

rosenkolev commented on src/MentorBot.WebLearningCenter/ClientApp/src/app/questions/components/question-page/question-page.component.ts:
```diff
@@ -0,0 +1,435 @@
+    itemContent: string,
+    parentsElement: any,
+    acquireTraitsElement: any,
+    requiredTraitsElement: any,
```
remove any type.

rosenkolev commented on src/MentorBot.WebLearningCenter/ClientApp/src/app/questions/components/question-page/question-page.component.ts:
```diff
@@ -0,0 +1,435 @@
+  editItem(node: TodoItemFlatNode) {
+    this.editedNode = node;
+    var addTraits = document.getElementsByClassName('add-trait');
+    for (var i = 0; i < addTraits.length; i++) {
```
this.editedNode.requiredTraits.push(node);

rosenkolev commented on src/MentorBot.WebLearningCenter/ClientApp/src/app/questions/components/question-page/question-page.component.ts:
```diff
@@ -0,0 +1,435 @@
+    var childNode = document.createElement('span');
+    childNode.className = 'badge badge-success add-trait';
+    childNode.textContent = e.toElement.textContent;
+    this.dragAndDropElement.appendChild(childNode);
```
directive
Renderer2

rosenkolev commented on src/MentorBot.WebLearningCenter/ClientApp/src/app/questions/components/question-page/question-page.component.ts:
```diff
@@ -0,0 +1,435 @@
+  }
+
+  prepareForDelete(value: string, element: any) {
+    var list = element.childNodes as Array<any>;
```
editedNode.fordelete.push(5)

rosenkolev commented on src/MentorBot.WebLearningCenter/ClientApp/src/app/questions/question.module.ts:
```diff
@@ -0,0 +1,105 @@
+import { MatTableModule } from '@angular/material/table';
+import { MatTabsModule } from '@angular/material/tabs';
+import { MatToolbarModule } from '@angular/material/toolbar';
+import { MatTooltipModule } from '@angular/material/tooltip';
```
remove not used

rosenkolev commented on src/MentorBot.WebLearningCenter/ClientApp/src/app/questions/question.service.ts:
```diff
@@ -0,0 +1,28 @@
+
+  static readonly deleteQuestion = 'delete-question';
+
+  constructor(private http: HttpClient) {}
```
private `readonly`

rosenkolev commented on src/MentorBot.WebLearningCenter/ClientApp/src/app/shared/confirmation/confirmation.component.html:
```diff
@@ -0,0 +1,26 @@
+<div class="backdrop"></div>
```
use angular material

Marin988 committed:
Remove userFlow plugin
Add question Factory for transforming questions to view models
Remove isEdited property from question view model and ordering the questions in the frontend
Break the code into smaller functions in "save questions", "delete question" and "get questions" and user flow processor

Marin988 committed:
Split the question component to many components
Remove unnessecary usings from question module
Take care of subscriptions

Marin988 committed:
Remove console logs

Marin988 committed:
Fix commit errors

Marin988 committed:
Fix queriesTest to expect 10 plugins instead of 11

Marin988 committed:
delete button doesn't show when we add a new node
drag and drop now only adds parent if we have dropped the parent in the parent section as opposed to just have dragged it over the section

Marin988 committed:
Add tests for saving questions and getting quesitons

Marin988 committed:
Specify the files in .editconfig which the rules would apply for (excluding html)
Change confirmation component's .css to .scss

Marin988 committed:
Fix whitespacing

Marin988 committed:
Add angular confirmation dialog and remove the previously used one

Marin988 committed:
Remove ShownQuestionsId property from State
Add userflow processor tests

rosenkolev commented on src/MentorBot.WebLearningCenter/ClientApp/src/app/questions/check-list-database.ts:
```diff
@@ -0,0 +1,105 @@
+  }
+
+  initialize() {
+    this._questionService.getQuestions().subscribe(questions => this.dataChange.next(questions));
```
never unsubscribe

rosenkolev commented on src/MentorBot.WebLearningCenter/ClientApp/src/app/questions/check-list-database.ts:
```diff
@@ -0,0 +1,105 @@
+      node.parents = parents;
+    }
+
+    let allQuestions = this.flattenTreeStructure(JSON.parse(JSON.stringify(this.dataChange.value)));
```
const

rosenkolev commented on src/MentorBot.WebLearningCenter/ClientApp/src/app/questions/check-list-database.ts:
```diff
@@ -0,0 +1,105 @@
+    }
+  }
+
+  flattenTreeStructure(questions: Question[]): Question[] {
```
```
flattenTreeStructure(questions: Question[]): Question[] {
  return flattenTreeStructure.reduce((arr, q) => {
     arr.push(q);
     arr.push.apply(arr, flattenTreeStructure(q.subQuestions));
  }, []);
}
```

rosenkolev commented on src/MentorBot.WebLearningCenter/ClientApp/src/app/questions/components/edit-question/edit-question.component.ts:
```diff
@@ -0,0 +1,69 @@
+  @Input() nodeExists?: boolean;
+  @Input() editedNode?: TodoItemFlatNode;
+  @Input() saveButtonIsNotValid?: boolean;
+  @Output() deleteTraitEvent = new EventEmitter<{ name: string; type: string }>();
```
@Output() traitEvent = new EventEmitter<{ name: string, type: string, eventType: Events }>()
enum EventTypes {
 AddRequired, Add
}

@Output() questionChanged = new EventEmitter<TodoItemFlatNode>();
updateTitle(title: string) {
  this.questionChanged .emit({ ...this.node, title });
}

rosenkolev commented on src/MentorBot.WebLearningCenter/ClientApp/src/app/questions/components/question-page/question-page.component.ts:
```diff
@@ -0,0 +1,241 @@
+  }
+
+  updateTitle({ title, isNotValid }: { title: string; isNotValid?: boolean }) {
+    (this.editedNode as TodoItemFlatNode).item = title;
```
this.editedNode = { ...this.editedNode, item: title }

rosenkolev commented on src/MentorBot.WebLearningCenter/ClientApp/src/app/questions/components/question-page/question-page.component.ts:
```diff
@@ -0,0 +1,241 @@
+  }
+
+  cancelEdit() {
+    this.editedNode = undefined;
```
cancelEdit() { this.resetData(); }

resetData(): void {
}

rosenkolev commented on src/MentorBot.WebLearningCenter/ClientApp/src/app/questions/components/question-page/question-page.component.ts:
```diff
@@ -0,0 +1,241 @@
+        message: 'Are you sure, you want to remove an employee: ' + this.editedNode?.item,
+      },
+    });
+    confirmDialog.afterClosed().subscribe(result => {
```
no unsubscribe

rosenkolev commented on src/MentorBot.WebLearningCenter/ClientApp/src/app/questions/components/edit-question/edit-question.component.ts:
```diff
@@ -0,0 +1,69 @@
+  }
+
+  changeEditNodeIsAnswer({ type, isNotValid }: { type: string; isNotValid?: boolean }) {
+    this.typeEvent.emit({ type, isNotValid });
```
this.typeEvent?.emit();

rosenkolev commented on src/MentorBot.WebLearningCenter/ClientApp/src/app/app.module.ts:
```diff
@@ -5,9 +5,17 @@ import { HttpClientModule, HTTP_INTERCEPTORS } from '@angular/common/http';
+    HttpClientModule,
+    BrowserAnimationsModule,
+    AppRoutingModule,
+    MatDialogModule,
```
remove MatDialogModule

rosenkolev commented on src/MentorBot.WebLearningCenter/ClientApp/src/app/questions/components/question-characteristics/question-characteristics.component.ts:
```diff
@@ -0,0 +1,32 @@
+  @Output() typeEvent = new EventEmitter<{ type: string; isNotValid: boolean | undefined }>();
+  @Output() titleEvent = new EventEmitter<{ title: string; isNotValid: boolean | undefined }>();
+  @Output() contentEvent = new EventEmitter<{ content: string; isNotValid: boolean | undefined }>();
+
```
public nodeType = NodeType;
```
<option [value]="nodeType.Aswer" [selected]="editedNode?.isAnswer" />
<option [value]="nodeType.Question" [selected]="!editedNode?.isAnswer" />
```

rosenkolev commented on src/MentorBot.WebLearningCenter/ClientApp/angular.json:
```diff
@@ -27,7 +27,10 @@
             "assets": ["src/assets"],
-            "styles": ["./node_modules/bootstrap-css-only/css/bootstrap.min.css"],
+            "styles": [
+              "./node_modules/bootstrap-css-only/css/bootstrap.min.css",
```
remove bootstrap

rosenkolev commented on src/MentorBot.WebLearningCenter/ClientApp/src/app/shared/confirm-dialog/components/confirm-dialog.component.ts:
```diff
@@ -0,0 +1,14 @@
+  message?: string;
+  constructor(public dialogRef: MatDialogRef<ConfirmDialogComponent>, @Inject(MAT_DIALOG_DATA) public data: any) {}
+
+  ngOnInit() {}
```
remove this

Marin988 committed:
Reduce the amount of Outputs
Create new models instead of using anonymous objects

rosenkolev commented on src/MentorBot.WebLearningCenter/ClientApp/src/app/questions/check-list-database.ts:
```diff
@@ -0,0 +1,80 @@
+
+/**
+ * Checklist database, it can build a tree structured Json object.
+ * Each node in Json object represents a to-do item or a category.
```
Update comment

rosenkolev commented on src/MentorBot.WebLearningCenter/ClientApp/src/app/questions/check-list-database.ts:
```diff
@@ -0,0 +1,80 @@
+    content?: string,
+    acquireTraits?: string[],
+    requiredTraits?: string[],
+    parents?: {
```
Record<string, string>
{ [key: string]: string }

rosenkolev commented on src/MentorBot.WebLearningCenter/ClientApp/src/app/questions/check-list-database.ts:
```diff
@@ -0,0 +1,80 @@
+      node.parents = parents;
+    }
+
+    let allQuestions = this.flattenTreeStructure([...this.dataChange.value]);
```
const

rosenkolev commented on src/MentorBot.WebLearningCenter/ClientApp/src/app/questions/check-list-database.ts:
```diff
@@ -0,0 +1,80 @@
+
+    let allQuestions = this.flattenTreeStructure([...this.dataChange.value]);
+
+    if (!allQuestions.find(n => n.title === name)) {
```
```
findElement(questions: Question[], title: string): Question {
  return questions.find(q => {
    if (q.title === title) {
      return q;
    }
    
    return findElement(q.subQuestions, title);
  })
}
```

rosenkolev commented on src/MentorBot.WebLearningCenter/ClientApp/src/app/questions/components/edit-question/edit-question.component.html:
```diff
@@ -0,0 +1,41 @@
+<div class="edit-content">
+    <div>
+        <div class="top-spacing">
+            <app-question-traits
```
empty

Marin988 committed:
Properly flattening the tree
Using enums instead of strings
Other fixes based on comments

rosenkolev commented on src/MentorBot.WebLearningCenter/ClientApp/src/app/shared/confirm-dialog/components/confirm-dialog.component.ts:
```diff
@@ -0,0 +1,11 @@
+export class ConfirmDialogComponent {
+  title?: string;
+  message?: string;
+  constructor(public dialogRef: MatDialogRef<ConfirmDialogComponent>, @Inject(MAT_DIALOG_DATA) public data: any) {}
```
do not use `any` type

rosenkolev commented on src/MentorBot.WebLearningCenter/ClientApp/src/app/questions/components/question-traits/question-traits.component.ts:
```diff
@@ -0,0 +1,29 @@
+
+  addTrait(input: HTMLInputElement): void {
+    this.traitEvent?.emit({ name: input.value, type: this.type, actionType: ActionType.Add });
+    input.value = '';
```
this is repeated down in `resetTraitFieldInput`. do
```
emitChangeEvent(name: string, actionType: ActionType): void {
   this.traitEvent...
}
addTrait(input: HTMLInputElement): void {
  this.emitChangeEvent(input.value, ActionType.Add);
  this.resetTraitFieldInput(input);
}

```

rosenkolev commented on src/MentorBot.WebLearningCenter/ClientApp/src/app/questions/components/question-page/question-page.component.ts:
```diff
@@ -0,0 +1,238 @@
+  onConfirm(): void {
+    const question = this.flatNodeMap.get(this.savingNode as TodoItemFlatNode);
+
+    this._questionService
```
duplicated code with save()

rosenkolev commented on src/MentorBot.WebLearningCenter/ClientApp/src/app/questions/components/question-page/question-page.component.ts:
```diff
@@ -0,0 +1,238 @@
+  }
+
+  updateNode({ title, isAnswer, content, isNotValid }: QuestionPropertiesChange) {
+    if (isAnswer !== undefined) {
```
(this.editedNode as TodoItemFlatNode) is repeating

rosenkolev commented on src/MentorBot.WebLearningCenter/ClientApp/src/app/questions/components/question-page/question-page.component.ts:
```diff
@@ -0,0 +1,238 @@
+  }
+
+  traitAction({ name, type, actionType }: TraitAction): void {
+    if (type === TraitTypes.Acquire && actionType === ActionType.Delete) {
```
Too big cyclomatic complexity.

rosenkolev commented on src/MentorBot.WebLearningCenter/ClientApp/src/app/questions/components/question-page/question-page.component.ts:
```diff
@@ -0,0 +1,238 @@
+  addParent?: boolean;
+  nodeExists?: boolean;
+
+  public nodeType = Object.values(NodeType).filter(value => typeof value === 'string');
```
Not used

rosenkolev commented on src/MentorBot.WebLearningCenter/ClientApp/src/app/questions/check-list-database.ts:
```diff
@@ -0,0 +1,101 @@
+  }
+
+  getFlatTree(): Question[] {
+    const flatTree = (nestedObjects: Question[]) => {
```
No single responsibility here.
1 function should do 1 thing

Marin988 committed:
Remove bootstrap
Use Angular Material instead

rosenkolev commented on src/MentorBot.WebLearningCenter/ClientApp/src/app/questions/components/question-characteristics/question-characteristics.component.ts:
```diff
@@ -0,0 +1,34 @@
+  public nodeType = NodeType;
+
+  changeEditNodeIsAnswer(type: number, isNotValid: boolean | undefined): void {
+    (this.editedNode as TodoItemFlatNode).isAnswer = type.toString() === this.nodeType.Answer.toString();
```
casts

rosenkolev commented on src/MentorBot.WebLearningCenter/ClientApp/src/app/questions/check-list-database.ts:
```diff
@@ -0,0 +1,101 @@
+      }, Object.create(null));
+    };
+
+    const allQuestions = flatTree([...this.dataChange.value]);
```
```
interface data {
  [childPropertyName]: data[];
  [prop: string]: unknow;
}
flatTree<T extends data>(nestedObjects: T[], childPropertyName: string): T[] => {
  return nestedObjects.reduce(function (resultArray: Question[], nestedObject: Question): Question[] {
        resultArray.push(nestedObject);
        const children = nestedObject[childPropertyName];
        if (children && children .length != 0) {
          Array.prototype.push(resultArray, flatTree(children, childPropertyName));
        }

        return resultArray;
      }, []);
};
```

rosenkolev commented on src/MentorBot.WebLearningCenter/ClientApp/src/app/questions/components/question-characteristics/question-characteristics.component.html:
```diff
@@ -0,0 +1,37 @@
+    </mat-select>
+</div>
+<div class="top-spacing">
+    {{editedNode?.item === undefined || '' ? '' : 'Edit'}} Title
```
`<span *ngIf="!editedNode?.item">Edit</span> Title`

rosenkolev commented on src/MentorBot.WebLearningCenter/ClientApp/src/app/questions/components/question-characteristics/question-characteristics.component.html:
```diff
@@ -0,0 +1,37 @@
+    <div>
+        <textarea class="textarea-size" [ngModel]="editedNode?.content" matInput #contentValue="ngModel" type="text"
+                  [disabled]="!editedNode?.isAnswer"
+                  [value]="!editedNode?.isAnswer ? '' : editedNode?.content === undefined || editedNode?.content === 'undefined' ? '' : editedNode?.content"
```
Too much checks

rosenkolev commented on src/MentorBot.WebLearningCenter/ClientApp/src/app/questions/components/question-characteristics/question-characteristics.component.ts:
```diff
@@ -0,0 +1,34 @@
+  styleUrls: ['question-characteristics.component.scss'],
+})
+export class QuestionCharacteristicsComponent {
+  @Input() editedNode?: TodoItemFlatNode;
```
```answerContent = '';

class s implement OnChange {
ngOnChange(changes: Changes): void {
 if (changes.editedNode) {
   this.answerContent  = this.editedNode?.content...
 }
}
```

Marin988 committed:
flatTree function now works for models extending interface data
Split functions into separate smaller ones
Manage execution of trait actions using an object instead of if statements

rosenkolev commented on src/MentorBot.WebLearningCenter/ClientApp/src/app/shared/utilities.service.ts:
```diff
@@ -0,0 +1,17 @@
+import { Injectable } from '@angular/core';
+import { data } from './utilities.models';
+
+@Injectable()
```
You can just do:
```
export function flatTree<T extends data>(nestedObjects: T[], childPropertyName: string): T[] {
  ...
  }
```

but that works too

Marin988 committed:
Swap mdb badges with angular chips
Remove any type from data interface

Marin988 committed:
Remove angular-bootstrap-md from package.json file
Add variable names to .cspell.json file to exclude them from spell check

Marin988 committed:
Change height of left side element in questions to fit the buttons better
Change the color of the nav menu
